tutor/utils.py

```diff
@@ -187,14 +188,15 @@ def is_a_tty() -> bool:
     Return True if stdin is able to allocate a tty. Tty allocation sometimes cannot be
     enabled, for instance in cron jobs
     """
-    return os.isatty(sys.stdin.fileno())
+    return sys.stdin.isatty()
 
 
 def execute(*command: str) -> int:
     click.echo(fmt.command(" ".join(command)))
     with subprocess.Popen(command) as p:
         try:
-            result = p.wait(timeout=None)
+            p.communicate()
+            result = p.returncode
```

Comment:
This change is making me nervous. I know that I fiddled quite a bit to get it just right. Are you positive that it's not breaking anything? Is there a particular reason you made this change?

Reply:
Hi @regisb , I particularly changed this because I need to mock one of the methods in the `subprocess.Popen` module (i.e `wait()`) and then inject mock functions to [simulate Exception or KeyboardInterrupt in test_utils.py](https://github.com/takato1314/tutor/blob/3e61feced4b6b126bddd631e29ba5782fe4a6489/tests/test_utils.py#L130). 

However, this is not possible with `wait()` because you are killing the sub-process and gracefully waiting it to shutdown [using the `wait()`](https://github.com/takato1314/tutor/blob/master/tutor/utils.py#L202) as well. So, it's not possible for me to mock the `wait()` method and hence I used the `communicate()` instead and mock that. Afaik, the `communicate()` should also able to run the sub-process and exists gracefully with the tuple returned.
![image](https://user-images.githubusercontent.com/10910968/144875303-c4cafb16-775b-4775-bff3-7cf6465f559b.png)

So far I have tested to run the tutor installation using `tutor local quickstart`, `tutor dev quickstart` and `tutor k8s quickstart` and installed tutor successfully. 

There's already the check at line 210 to check for the fail result and throw error. If you wanna to be extra safe, maybe we could also check the tuple results from the `p.communicate()`?

```
output, error = p.communicate()

## Check error if exist?
if result > 0 or error:
            raise exceptions.TutorError(
                "Command failed with status {}: {}".format(result, " ".join(command))
            )
```

Reply:
Hi @regisb , fyi idm to revert this part of change as well if you are still worry about it and maybe we can address it in different PR.

Reply:
Hi @regisb , i reverted thiis part to use the original functions and add in unit tests based on those functions via my repo PR  [#8](https://github.com/takato1314/tutor/pull/8).